From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Tue, 22 Sep 2020 01:49:19 -0700
Subject: [PATCH] Optimise non-flush packet sending

Original license: GPLv3
Original project: https://github.com/PaperMC/Paper
Paper pull request: https://github.com/PaperMC/Paper/pull/10172

Places like entity tracking make heavy use of packet sending,
and internally netty will use some very expensive thread wakeup
calls when scheduling.

Thanks to various hacks in ProtocolLib as well as other
plugins, we cannot simply use a queue of packets to group
send on execute. We have to call execute for each packet.

Tux's suggestion here is exactly what was needed - tag
the Runnable indicating it should not make a wakeup call.

Big thanks to Tux for making this possible as I had given
up on this optimisation before he came along.

Locally this patch drops the entity tracker tick by a full 1.5x.

Co-authored-by: Quang Tran <3d7777456@gmail.com>

diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 21e891ecdb90a1828428441aef9c2593253751f8..c109fafe09f144079d6cba1325baec1602beb8a0 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -125,6 +125,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public org.bukkit.event.player.PlayerResourcePackStatusEvent.@Nullable Status resourcePackStatus; // Paper
     // Paper start - Optimize network
     public boolean isPending = true;
+    private io.netty.channel.SingleThreadEventLoop eventLoop; // Paper - optimise packets that are not flushed
     public boolean queueImmunity;
     // Paper end - Optimize network
 
@@ -136,6 +137,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public void channelActive(ChannelHandlerContext ctx) throws Exception {
         super.channelActive(ctx);
         this.channel = ctx.channel();
+        this.eventLoop = (io.netty.channel.SingleThreadEventLoop) this.channel.eventLoop(); // Paper - optimise packets that are not flushed
         this.address = this.channel.remoteAddress();
         this.preparing = false; // Spigot
         if (this.delayedDisconnect != null) {
@@ -448,6 +450,11 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.channel.eventLoop().inEventLoop()) {
             this.doSendPacket(packet, sendListener, flush);
         } else {
+            // Paper start - optimise packets that are not flushed
+            if (!flush && org.dreeam.leaf.config.modules.network.OptimizeNonFlushPacketSending.enabled) {
+                this.eventLoop.lazyExecute(() -> this.doSendPacket(packet, sendListener, false));
+            } else
+                // Paper end - optimise packets that are not flushed
             this.channel.eventLoop().execute(() -> this.doSendPacket(packet, sendListener, flush));
         }
     }
